---
title: "Quality Assessment and Trimming Loggerhead Data"
author: "Jordan"
date: "2025-03-04"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.align = "center", fig.path = "../figures/01_QualityTrimming")
```

# Goals of this file 

1. Use raw fastq files and generate quality plots to assess quality of reads.
2. Filter and trim out bad sequences and bases from our sequencing files. 
3. Write out fastq files with high quality sequences. 
4. Evaluate the quality from our filter and trim. 
5. Infer Errors on forward and reverse reads individually.
6. Identified ASVs on forward and reverse reads separately, using the error model.  
7. Merge forward and reverse ASVs into "contiguous ASVs".  
8. Generate the ASV count table. (`otu_table` input for phyloseq.). 

Output that we will create in this file: 

1. Quality plots to asses pre- and post-quality trimming. 
2. Written files that include the filtered and trimmed data. (Note that this will be written during the `filterAndTrim()` function.)

# Setting up the environment

### Set the seed 

Why is it useful to set a seed? 

1. *Reproducibility*:
    - If you’re using random number generation (e.g., bootstrapping, simulations, machine learning models), setting a seed ensures that you (and others) can get the same results every time you run the code.
2. *Debugging & Collaboration*:
    - If your analysis involves randomness (e.g., shuffling datasets, training ML models), setting a seed makes it easier to debug and compare results across different runs or machines.
    - Colleagues can run your code and get identical results, making scientific collaboration more reliable.
3. *Benchmarking & Model Comparisons*:
    - In machine learning, setting a seed ensures that models trained with random splits of data or initial weights are comparable across different trials.
    - If you don’t set a seed, each run might give slightly different performance metrics, making it harder to evaluate true differences between models.
4. *Avoiding Accidental Bias in Random Sampling*:
    - When selecting random subsets of data (e.g., cross-validation splits), setting a seed ensures that all experiments are based on the same random selection, reducing unwanted variability.

```{r set-seed}
set.seed(238428)
```

### Timing of Script
```{r rmd-start}
# What time did we start running this script? 
start_time <- Sys.time()
start_time
```

# Load Libraries 
```{r load-libraries}
# Efficient package loading with pacman 
# Don't forget to install pacman and DT if you don't have it yet. :)
pacman::p_load(tidyverse, dada2, phyloseq, patchwork, DT, devtools, install = FALSE)
```
# Reading Raw Sequencing Files 
```{r load-data}
# Set the raw fastq path to the raw sequencing files 
# Path to the fastq files 
raw_fastqs_path <- "./data/01_DADA2/01_raw_gzipped_fastqs"
raw_fastqs_path

# What files are in this path? Intuition Check 
head(list.files(raw_fastqs_path))

# How many files are there?
length(list.files(raw_fastqs_path))

# Create vector of forward reads
forward_reads <- list.files(raw_fastqs_path, pattern = "R1_001.fastq.gz", full.names = TRUE)

# Intuition Checks 
head(forward_reads)
# Intuition check #2: We should have fewer reads in the forward vector than in total 
stopifnot(length(forward_reads) < length(list.files(raw_fastqs_path)))

# Create a vector of reverse reads 
reverse_reads <- list.files(raw_fastqs_path, pattern = "R2_001.fastq.gz", full.names = TRUE)

# Intuition Checks
head(reverse_reads)
# Intuition check #2: Need to have equal number of forward and reverse files 
stopifnot(length(reverse_reads) == length(forward_reads))
```

# Assess Raw Read Quality 

## Evaluate raw sequence quality 

Let's see the quality of the raw reads *before* we trim

## Plot 12 random samples of plots
```{r raw-quality-plot, fig.width=12, fig.height=8}
# Randomly select 12 samples from dataset to evaluate 
# Selecting 12 is typically better than 2 (like we did in class for efficiency)
random_samples <- sample(1:length(reverse_reads), size = 12)
random_samples

# Calculate and plot quality of these two samples
forward_filteredQual_plot_12 <- plotQualityProfile(forward_reads[random_samples]) + 
  guides(fill = "none") +
  labs(title = "Forward Read: Raw Quality")

reverse_filteredQual_plot_12 <- plotQualityProfile(reverse_reads[random_samples]) + 
  guides(fill = "none") +
  labs(title = "Reverse Read: Raw Quality")

# Plot them together with patchwork
forward_filteredQual_plot_12 + reverse_filteredQual_plot_12

```
**Acceptable Phred Scores for ASVs with DADA2**

In DADA2, quality filtering is crucial for obtaining reliable biological sequences. Below are the acceptable **Phred score** thresholds for DADA2 preprocessing:

| **Q Score**   | **Error Probability**    | **DADA2 Interpretation**            |
|--------------|------------------------|------------------------------------|
| **40**        | 0.0001 (1 in 10,000) | Very high-quality , dream come true   | 
| **≥30**     | ≤ 0.001 (1 in 1,000) | Excellent quality, keep                |
| **25-30** | ~0.003 - 0.01        | Can be acceptable, but more important to trim low-quality tails |
| **20-25** | ~0.01 - 0.03         | Lower quality, only acceptable if read length is absolutely critical |
| **<20**    | > 0.01 (1 in 100)    | Poor quality, remove                    |
| **<10**     | >0.1 (1 in 10)       | Very unreliable, absolutely remove      |
---

Some summary take-aways:

- **30 is good** – Bases with Q30+ scores have a **99.9% accuracy rate**, making them highly reliable.  
- **25-30 can be acceptable**, but trimming is preferred to remove low-quality tails.  
- **20-25** should only be kept if read length is absolutely critical (*e.g.* to ensure forward & reverse read overlap)
- **<20 discarded** – Bases with Phred scores below 20 have an **error rate >1%** and should 100% be **trimmed from the dataset**.  

Remember, every base counts when we assign ASVs, they are called **sequence variants** for a reason! And, we want to be sure. :)

---
## Aggregated Raw Quality Plots 

Next, we will plot all of the samples aggregated into one forward (left) and one reverse read (right) plot. 

```{r raw-aggregate-plot, fig.width=5.5, fig.height=3.5}
# Aggregate all QC plots 
# Forward reads
forward_preQC_plot <- 
  plotQualityProfile(forward_reads, aggregate = TRUE) + 
  labs(title = "Forward Pre-QC")

# reverse reads
reverse_preQC_plot <- 
  plotQualityProfile(reverse_reads, aggregate = TRUE) + 
  labs(title = "Reverse Pre-QC")

# Now, let's put the two plots together
preQC_aggregate_plot <- 
  # Plot the forward and reverse together 
  forward_preQC_plot + reverse_preQC_plot
# Show the plot
preQC_aggregate_plot

```

Above are plots showing the pre-QC quality scores of the raw sequencing data for the project. We can draw a few conclusions from the plots above, which were generated with 2x300 bp sequencing on an Illumina MiSeq v3 Sequencer: 

1.	*High Quality in Early and Mid-Reads (~0-175 Cycles/Bases):* 
    - Quality is above Q30 in the first half of both forward and reverse reads.
    - Notably, the forward reads continue with high quality scores farther than the reverse reads (~0-250 Cycles/Bases).
    - This suggests that at least half of the read length is high-quality and can be retained, and that we don't need to trim the beginning of the reads.
2.	*Decline in Forward Reads After ~250 Cycles/Bases:*
    - The last ~ 8 bases of the forward reads shows a drop below 30, indicating a high sequencing error rate.
        - This was also seen in fastQC and multiQC, but only the last 5 bases look <25.
    - **Solution**: We will truncate the read using ` truncLen = 295` to remove those final 5 poor-quality bases at the end of the read. The beginning of the read is >30 so we are ok at the beginning of the read.
3.  *Decline in Reverse Reads After ~175 Cycles/Bases:*
    - The reverse reads degrade more quickly than the forward reads, more on this below.
    - The last ~ 70 bases of the reverse reads shows a drop below 30, indicating a high sequencing error rate.
      - This was also seen in fastQC and multiQC, but only the last ~50 bases look <25.
    - **Solution**: We will truncate the read using ` truncLen = 250` to remove those final 50 poor-quality bases at the end of the read. The beginning of the read is >30 so we are ok at the beginning of the read.
4.	*Forward Reads Maintain Quality Slightly Longer*
    - Forward reads degrade after ~250 cycles, while reverse reads degrade after ~175 cycles.
    
General notes on Illumina sequencing

- **Forward & Reverse Reads:** 
  - Typically all reads decline in quality over the sequencing run due to phasing (*i.e.* when the clusters become out of sync with each other later in the run). 
  - Read quality at the beginning of the read can be common due to primer/adapter sequence artifacts
- **Reverse Reads Quality is generally lower because:** 
  - Sequencing of the reverse reads happens second, after chemistry degrades a bit after sequencing the forward reads. 
  - Physical flipping of the sequence is required.
  - Phasing errors are enhanced, more accumulated errors.
  - Optical strength is weaker compared to the first sequencing, leading to more miscalls. 

####  Prepare a placeholder for filtered reads 
```{r prep-filtered-sequences}
# Create vector of sample names from the filenames 
sample_names <- sapply(strsplit(basename(forward_reads), "_"), `[`,1)

# Intuition Check 
head(sample_names)

# Place filtered reads into filtered_fastqs_path
filtered_fastqs_path <- "data/01_DADA2/02_filtered_fastqs"

# Intuition Check 
filtered_fastqs_path

# create 2 vectors: filtered_forward_reads & filtered_reverse_reads
filtered_forward_reads <- 
  file.path(filtered_fastqs_path, paste0(sample_names, "_R1_filtered.fastq.gz"))

# Intuition Check 
length(filtered_forward_reads)

# reverse reads
filtered_reverse_reads <- 
  file.path(filtered_fastqs_path, paste0(sample_names, "_R2_filtered.fastq.gz"))

# Intuition Check 
head(filtered_reverse_reads)

```

# Filter and Trim Reads

```{r filter-and-trim}
# Assign and write out filtered fastq files 
# Paper noted that the obtained sequences had non-biological sequences trimmed by the sequencing facility. We do not need to trim out primers.
# I'm nervous and I want to be safe and use the default parameters

filtered_reads <- 
  filterAndTrim(fwd = forward_reads, filt = filtered_forward_reads,
              rev = reverse_reads, filt.rev = filtered_reverse_reads,
              maxN = 0, maxEE = c(2,2),
              truncQ = 2, rm.phix = TRUE, compress = TRUE, 
              # Please set threads to be 8-10 (we want to respect others using the server!)
              # Note that if TRUE, it will use ALL threads (making it hard for others to use the server)
              multithread = 8)

```

# Assess Trimmed Read Quality 

```{r filterTrim-quality-plots,  fig.width=12, fig.height=8}
# Plot the 12 random samples after QC
forward_filteredQual_plot_12 <- 
  plotQualityProfile(filtered_forward_reads[random_samples]) + 
  labs(title = "Trimmed Forward Read Quality")

reverse_filteredQual_plot_12 <- 
  plotQualityProfile(filtered_reverse_reads[random_samples]) + 
  labs(title = "Trimmed Reverse Read Quality")

# Put the two plots together 
forward_filteredQual_plot_12 + reverse_filteredQual_plot_12

```

## Aggregated Trimmed Plots

```{r qc-aggregate-plot, fig.width=5.5, fig.height=3.5}
# Forward reads
forward_postQC_plot <- 
  plotQualityProfile(filtered_forward_reads, aggregate = TRUE) + 
  labs(title = "Forward Post-QC")

# reverse reads
reverse_postQC_plot <- 
  plotQualityProfile(filtered_reverse_reads, aggregate = TRUE) + 
  labs(title = "Reverse Post-QC")

# Now, let's put the two plots together
postQC_aggregate_plot <- 
  # Plot the forward and reverse together 
  forward_postQC_plot + reverse_postQC_plot

# Show the plot
postQC_aggregate_plot
```

This post-QC quality score plot from 102 aggregated Illumina sequencing files shows the forward (left) and reverse (right) reads after quality filtering and trimming.

- *Forward Reads Post-QC*
  - High-quality (Q30+) bases throughout most of the read length.
  - A slight decline after ~250 cycles, only the last ~5 bases below a Phred of 30.
- *Reverse Reads Post-QC*
  - First half (~0-175 cycles) remain stable and high-quality (Q30+).
  - A noticeable decline in quality after ~175 cycles, improved compared to pre-QC. The last ~40 bases of post-filtered reads < Q30, compared the last ~70 bases of pre-filtered reads < Q30.

**Takeaway:** Quality filtering partially successfully retained high-quality bases while removing some low-quality ends.

To further confirm that things look ok, let's take a look at the read retention of our QC step.

## Read Retention Post-QC

```{r filterTrim-stats, message = FALSE, fig.height = 2.5, fig.width=8}
# Make output into dataframe 
filtered_df <- as.data.frame(filtered_reads) %>%
  mutate(percent.retained = reads.out/reads.in)

# Intuition check
# Visualize it in table format 
DT::datatable(filtered_df)

# Let's calculate some statistics
read_stats_df <- 
  filtered_df %>%
  reframe(median_reads_in = median(reads.in),
          median_reads_out = median(reads.out),
          median_percent_retained = (median(reads.out)/median(reads.in)),
          max_percent_retained = max(reads.out/reads.in),
          min_percent_retained = min(reads.out/reads.in))

# Take a look at it!
read_stats_df

# Plot it 
numSeqs_QC_dotplot <-
  filtered_df %>%
  ggplot(aes(x = reads.in, y = reads.out)) + 
  geom_point(alpha = 0.5, size = 2) + 
  labs(x = "# of Raw Seqs", 
       y = "# of Seqs Retained") + 
  # Now let's add a 1:1 line for reference of keeping 100% of the reads
  geom_abline(slope=1, intercept = 0, color = "deeppink")

# Now, let's look at the number of reads retained in a histogram
numRetained_QC_histplot <- 
  filtered_df %>%
  ggplot(aes(x = reads.out)) + 
  geom_histogram() + 
  labs(x = "# of Seqs Retained", 
       y = "# of Samples") 

# Create a histogram of percent reads retained in a histogram
percSeqs_QC_histplot <- 
  filtered_df %>%
  ggplot(aes(x = percent.retained)) + 
  geom_histogram() + 
  labs(x = "% of Seqs Retained", 
       y = "# of Samples") + 
  # Set the scale to be between 0-1 (0-100%)
  scale_x_continuous(limits = c(0, 1))

# Now, let's put the plots together
numSeqs_QC_dotplot + numRetained_QC_histplot + percSeqs_QC_histplot + 
  plot_annotation(tag_levels = 'A')

```

This figure presents three panels showing how many sequences were retained after quality filtering and trimming in the DADA2 pipeline. Let’s break down each panel:

**Panel A: Scatter Plot of Raw vs. Retained Sequences:**  

- X-axis: Number of raw sequences before filtering.
- Y-axis: Number of sequences retained after filtering.
- Pink Line: The diagonal line represents perfect retention (i.e., no sequences lost).

*Interpretation of Panel A:*  

Note: the x- and y-axis scales are different
- In general, most points are below the diagonal, meaning most samples lost sequences during filtering.
- As the number of raw sequences in a sample increases, the number of sequences retained seems to decrease.
- There are two outliers with a higher number of raw reads that fall far from the line.
Overall, retention of reads across most samples was not ideal.

**Panel B: Histogram of the Number of Sequences Retained per Sample**  

- X-axis: Number of sequences retained per sample.
- Y-axis: Number of samples with that many retained sequences.

*Interpretation of Panel B*  

- The majority of samples retained less than 300,000 sequences, but most had less than 500,000 raw reads.
- The two ouliers from above are biasing the scale. These two had ~ 1,180,000 raw reads and retained ~ 800,000 sequences.
- There are some samples that has almost no reads and will need to be removed!
Overall, it is difficult to tell whether samples retained a high number of sequences, and a graph of percent of sequences retained will provide better visual information (see Panel C).

**Panel C: Histogram of Percent of Sequences Retained**  

- X-axis: Proportion (%) of sequences retained per sample.
- Y-axis: Number of samples at each proportion.

*Interpretation of Panel C*. 

- Most samples retained >50% of their sequences, the largest group retaining just over 60%.
- Filtering did remove a lot of sequences.
- **Max % Retained** is `r read_stats_df$max_percent_retained` is ok while **min % retained is** `r read_stats_df$min_percent_retained` is probably really bad.
- A **median % retained ** of `r read_stats_df$median_percent_retained` is closer to the maximum value retained which is good, but the number itself is ok. 


**Consider re-running your `filterAndTrim()` if:** 

- If important samples lost too many reads, consider relaxing `maxEE` (expected errors) or adjusting truncation lengths (`truncLen`).
- Low merging success later on in the DADA2 workflow (suggests too much length variation).
- Reverse read degradation still affects error modeling (trim further if needed).

### Visualize QC differences in plot
```{r pre-post-QC-plot, fig.width=6, fig.height=5.5}
# Plot the pre and post together in one plot
preQC_aggregate_plot / postQC_aggregate_plot

```
**Quality Score Improvements**

- *Forward Reads (Pre vs. Post-QC)*
  - *Beginning of read*: The quality remians consistently high Q30-Q40. Not a large difference in pre- and post-QC. No bases were trimmed.
  - *Middle of read*: The quality continues to be high (>Q30), then drops slightly ~ 250 bp, but the post-QC reads retain higher quality scores in this region compared to pre-QC.
  - *End of read*: Both pre- and post-QC reads drop below Q30 at ~ 290 bp. These should probably be trimmed in further filtering steps.
- *Reverse Reads (Pre vs. Post-QC)*
  - *Beginning of read*: The quality of remians consistently high Q30-Q40. Not a large difference in pre- and post-QC. No bases were trimmed.
  - *Middle of read*: The quality starts to drop ~ 175 bp, but the post-QC reads retain higher quality scores in this region compared to pre-QC.
  - *End of read*: While both pre- and post-QC reads fall below Q30 well before the end of the cycle, post-QC falls below at ~ 260 bp while pre-QC falls below at ~ 230 bp. This is an improvement but additional bases should be trimmed in further filtering steps. 
  
Overall, I don't think I have sufficiently QC'd my sequencing data. There are still quite a bit of seqences that fall below Q30, especially with the end of reverse reads. However, I am nervous to trim off too many bases, as this will shorten the ASVs, but I think at least 10bp could work.
  
# Check Render Time
```{r stop-time}
# Take the time now that we are at the end of the script
end_time <- Sys.time()
end_time 

# Echo the elapsed time
elapsed_time <- round((end_time - start_time), 3)
elapsed_time
```
# Session Information 
```{rsession-info}
# Ensure reproducibility 
devtools::session_info()
```




